src/libraries/System.Security.Cryptography.X509Certificates/tests/PublicKeyTests.cs

```diff
@@ -558,6 +640,149 @@ public static void TestKey_BrainpoolP160r1()
+        {
+            using DSA dsa = DSA.Create();
+            dsa.ImportFromPem(TestData.DsaPkcs8PublicKey);
+            PublicKey key = new PublicKey(dsa);
+
+            Span<byte> algSpki = dsa.ExportSubjectPublicKeyInfo();
+            Assert.True(algSpki.SequenceEqual(key.ExportSubjectPublicKeyInfo()), "SequenceEquals(ExportSubjectPublicKeyInfo)");
+
+            // Just right
+            Assert.True(key.TryExportSubjectPublicKeyInfo(algSpki, out int written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(algSpki.Length, written);
+
+            // Too small
+            Assert.False(key.TryExportSubjectPublicKeyInfo(algSpki.Slice(1), out written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(0, written);
+        }
+
+        [Fact]
+        public static void ExportSubjectPublicKeyInfo_ECDSA()
+        {
+            using ECDsa ecdsa = ECDsa.Create();
+            ecdsa.ImportFromPem(TestData.ECDsaPkcs8PublicKey);
+            PublicKey key = new PublicKey(ecdsa);
+
+            Span<byte> algSpki = ecdsa.ExportSubjectPublicKeyInfo();
+            Assert.True(algSpki.SequenceEqual(key.ExportSubjectPublicKeyInfo()), "SequenceEquals(ExportSubjectPublicKeyInfo)");
+
+            // Just right
+            Assert.True(key.TryExportSubjectPublicKeyInfo(algSpki, out int written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(algSpki.Length, written);
+
+            // Too small
+            Assert.False(key.TryExportSubjectPublicKeyInfo(algSpki.Slice(1), out written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(0, written);
+        }
+
+        [Fact]
+        public static void ExportSubjectPublicKeyInfo_ECDH()
+        {
+            using ECDiffieHellman ecdh = ECDiffieHellman.Create();
+            ecdh.ImportFromPem(TestData.EcDhPkcs8Key);
+            PublicKey key = new PublicKey(ecdh);
+
+            Span<byte> algSpki = ecdh.ExportSubjectPublicKeyInfo();
+            Assert.True(algSpki.SequenceEqual(key.ExportSubjectPublicKeyInfo()), "SequenceEquals(ExportSubjectPublicKeyInfo)");
+
+            // Just right
+            Assert.True(key.TryExportSubjectPublicKeyInfo(algSpki, out int written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(algSpki.Length, written);
+
+            // Too small
+            Assert.False(key.TryExportSubjectPublicKeyInfo(algSpki.Slice(1), out written), nameof(key.TryExportSubjectPublicKeyInfo));
+            Assert.Equal(0, written);
+        }
+
+        [Fact]
+        public static void CreateFromSubjectPublicKeyInfo_Roundtrip_RSA()
+        {
+            using RSA rsa = RSA.Create();
+            rsa.ImportFromPem(TestData.RsaPkcs8PublicKey);
+            byte[] spki = rsa.ExportSubjectPublicKeyInfo();
+
+            PublicKey key = PublicKey.CreateFromSubjectPublicKeyInfo(spki, out int read);
+
+            Assert.True(key.Key is RSA, "is RSA");
```

Comment:
Something like `Assert.IsAssignableTo<RSA>` would lead to a more easily diagnosed failure.